File: foundations-jdbc-kotlin/src/kotlin/dev/typr/foundationskt/connect/ConnectionSource.kt
package dev.typr.foundationskt.connect

import dev.typr.foundationskt.Transactor
import java.sql.Connection

interface ConnectionSource {
    fun getConnection(): Connection
    fun transactor(): Transactor

    companion object {
        @JvmStatic
        fun of(config: DatabaseConfig): ConnectionSource {
            val java = dev.typr.foundations.connect.ConnectionSource.of(config)
            return wrap(java)
        }

        @JvmStatic
        fun of(config: DatabaseConfig, settings: ConnectionSettings): ConnectionSource {
            val java = dev.typr.foundations.connect.ConnectionSource.of(config, settings)
            return wrap(java)
        }

        internal fun wrap(java: dev.typr.foundations.connect.ConnectionSource): ConnectionSource =
            object : ConnectionSource {
                override fun getConnection(): Connection = java.getConnection()
                override fun transactor(): Transactor =
                    Transactor(java.transactor())
            }
    }
}

File: foundations-jdbc-scala/src/scala/dev/typr/foundationssc/ConnectionSource.scala
package dev.typr.foundationssc

import java.sql.Connection

trait ConnectionSource:
  def getConnection(): Connection
  def transactor(): Transactor = transactor(Transactor.defaultStrategy())
  def transactor(strategy: Transactor.Strategy): Transactor

object ConnectionSource:
  def of(config: connect.DatabaseConfig): ConnectionSource =
    val java = dev.typr.foundations.connect.ConnectionSource.of(config)
    wrap(java)

  def of(config: connect.DatabaseConfig, settings: connect.ConnectionSettings): ConnectionSource =
    val java = dev.typr.foundations.connect.ConnectionSource.of(config, settings)
    wrap(java)

  private[foundationssc] def wrap(java: dev.typr.foundations.connect.ConnectionSource): ConnectionSource =
    new ConnectionSource:
      override def getConnection(): Connection = java.getConnection()
      override def transactor(strategy: Transactor.Strategy): Transactor =
        Transactor(java.transactor(strategy))
